ddcommon/src/unix_utils.rs

```diff
@@ -170,3 +171,70 @@ pub fn wait_for_pollhup(target_fd: RawFd, timeout_ms: i32) -> anyhow::Result<boo
         }
     }
 }
+
+#[cfg(target_os = "macos")]
+pub fn alt_fork() -> i32 {
+    // There is a lower-level `__fork()` function in macOS, and we can call it from Rust, but the
+    // runtime is much stricter about which operations (e.g., no malloc) are allowed in the child.
+    // This somewhat defeats the purpose, so macOS for now will just have to live with atfork
+    // handlers.
+    unsafe { libc::fork() }
+}
+
+#[cfg(target_os = "linux")]
+fn is_being_traced() -> io::Result<bool> {
+    // Check to see whether we are being traced.  This will fail on systems where procfs is
+    // unavailable, but presumably in those systems `ptrace()` is also unavailable.
+    // The caller is free to treat a failure as a false.
+    let file = File::open("/proc/self/status")?;
+    let reader = BufReader::new(file);
+
+    for line in reader.lines() {
+        let line = line?;
+        if line.starts_with("TracerPid:") {
+            let tracer_pid = line.split_whitespace().nth(1).unwrap_or("0");
+            return Ok(tracer_pid != "0");
+        }
+    }
+
+    Ok(false)
+}
+
+#[cfg(target_os = "linux")]
+pub fn alt_fork() -> pid_t {
+    use libc::{
+        c_ulong, c_void, syscall, SYS_clone, CLONE_CHILD_CLEARTID, CLONE_CHILD_SETTID,
+        CLONE_PTRACE, SIGCHLD,
+    };
+
+    let mut _ptid: pid_t = 0;
+    let mut _ctid: pid_t = 0;
+
+    // Check whether we're traced before we fork.
+    let being_traced = is_being_traced().unwrap_or(false);
```

Comment:
Do we have a system test for this?

Reply:
(this handles the case where ptrace is attached to a process and it requires slightly different semantics here)